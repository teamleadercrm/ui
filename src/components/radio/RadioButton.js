import cx from 'classnames';
import omit from 'lodash.omit';
import PropTypes from 'prop-types';
import React, { createRef, PureComponent } from 'react';
import Box, { omitBoxProps, pickBoxProps } from '../box';
import { TextBodyCompact, TextDisplay, TextSmall } from '../typography';
import theme from './theme.css';

class RadioButton extends PureComponent {
  inputNode = createRef();

  handleToggle = (event) => {
    const { disabled, checked, onChange } = this.props;

    if (event.pageX !== 0 && event.pageY !== 0) {
      this.blur();
    }

    if (!disabled && onChange) {
      onChange(!checked, event);
    }
  };

  blur() {
    if (this.inputNode.current) {
      this.inputNode.current.blur();
    }
  }

  focus() {
    if (this.inputNode.current) {
      this.inputNode.current.focus();
    }
  }

  render() {
    const { checked, disabled, className, size, label, children, onMouseEnter, onMouseLeave, ...others } = this.props;
    const TextElement = size === 'small' ? TextSmall : size === 'medium' ? TextBodyCompact : TextDisplay;

    const restProps = omit(others, ['onChange']);
    const boxProps = pickBoxProps(restProps);
    const inputProps = omitBoxProps(restProps);

    const classNames = cx(
      theme['radiobutton'],
      theme[`is-${size}`],
      {
        [theme['is-checked']]: checked,
        [theme['is-disabled']]: disabled,
      },
      className,
    );

    return (
      <Box
        element="label"
        data-teamleader-ui="radiobutton"
        className={classNames}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        {...boxProps}
      >
        <input
          className={theme['input']}
          type="radio"
          checked={checked}
          disabled={disabled}
          onClick={this.handleToggle}
          readOnly
          ref={this.inputNode}
          {...inputProps}
        />
        <span className={theme['shape']} />
        {(label || children) && (
          <span className={theme['label']}>
            {label && (
              <TextElement element="span" color={disabled ? 'neutral' : 'teal'}>
                {label}
              </TextElement>
            )}
            {children}
          </span>
        )}
      </Box>
    );
  }
}

RadioButton.propTypes = {
  checked: PropTypes.bool,
  children: PropTypes.node,
  disabled: PropTypes.bool,
  name: PropTypes.string,
  className: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

RadioButton.defaultProps = {
  checked: false,
  disabled: false,
  size: 'medium',
};

export default RadioButton;
