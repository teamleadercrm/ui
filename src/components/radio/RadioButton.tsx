import cx from 'classnames';
import omit from 'lodash.omit';
import React, { useRef, MouseEvent, ReactNode } from 'react';
import { GenericComponent } from '../../@types/types';
import Box, { omitBoxProps, pickBoxProps } from '../box';
import { TextBodyCompact, TextDisplay, TextSmall } from '../typography';
import { SIZES } from '../../constants';
import { BoxProps } from '../box/Box';
import theme from './theme.css';

export interface RadioButtonProps extends Omit<BoxProps, 'className' | 'children'> {
  checked?: boolean;
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  name?: string;
  onChange?: (checked: boolean, event: MouseEvent) => void;
  onMouseEnter?: (event: MouseEvent) => void;
  onMouseLeave?: (event: MouseEvent) => void;
  label?: string;
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest' | 'tiny' | 'hero'>;
  value: string | boolean;
}

const RadioButton: GenericComponent<RadioButtonProps> = ({
  checked = false,
  children,
  className,
  disabled = false,
  name,
  onChange,
  onMouseEnter,
  onMouseLeave,
  label,
  size = 'medium',
  ...others
}) => {
  const inputNode = useRef<HTMLInputElement | null>(null);
  const TextElement = size === 'small' ? TextSmall : size === 'medium' ? TextBodyCompact : TextDisplay;

  const restProps = omit(others, ['onChange']);
  const boxProps = pickBoxProps(restProps);
  const inputProps = omitBoxProps(restProps);

  const classNames = cx(
    theme['radiobutton'],
    theme[`is-${size}`],
    {
      [theme['is-checked']]: checked,
      [theme['is-disabled']]: disabled,
    },
    className,
  );

  const blur = () => {
    if (inputNode.current) {
      inputNode.current.blur();
    }
  };

  const handleToggle = (event: React.MouseEvent) => {
    if (event.pageX !== 0 && event.pageY !== 0) {
      blur();
    }

    if (!disabled && onChange) {
      onChange(!checked, event);
    }
  };

  return (
    <Box
      element="label"
      data-teamleader-ui="radiobutton"
      className={classNames}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      {...boxProps}
    >
      <input
        className={theme['input']}
        type="radio"
        checked={checked}
        disabled={disabled}
        onClick={handleToggle}
        readOnly
        ref={inputNode}
        name={name}
        {...inputProps}
      />
      <span className={theme['shape']} />
      {(label || children) && (
        <span className={theme['label']}>
          {label && (
            <TextElement element="span" color={disabled ? 'neutral' : 'teal'}>
              {label}
            </TextElement>
          )}
          {children}
        </span>
      )}
    </Box>
  );
};

export default RadioButton;
