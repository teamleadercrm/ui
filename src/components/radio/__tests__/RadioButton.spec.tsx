import React from 'react';
import { render } from '@testing-library/react';
import RadioButton from '../RadioButton';

describe('Component - RadioButton', () => {
  it('renders', async () => {
    const { asFragment } = render(<RadioButton value="foobar" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
