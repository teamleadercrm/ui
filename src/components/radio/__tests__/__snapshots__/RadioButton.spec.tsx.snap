// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - RadioButton renders 1`] = `
<DocumentFragment>
  <label
    class="box radiobutton is-medium"
    data-teamleader-ui="radiobutton"
  >
    <input
      class="input"
      readonly=""
      type="radio"
      value="foobar"
    />
    <span
      class="shape"
    />
  </label>
</DocumentFragment>
`;
